zboxcore/sdk/registry_file_manager.go

```diff
@@ -0,0 +1,203 @@
+
+type registryFileManager interface {
+	Update(data []byte) error
+	Get() ([]byte, common.Timestamp, error)
+	GetLastUpdateTimestamp() (common.Timestamp, error)
+}
+
+// registryFileStore manages the storing of registry file on allocation.
+type registryFileStore struct {
+	registryFilePath string
+	allocation       *Allocation
+	fileStorer       allocationFileStorer
+}
+
+var _ registryFileManager = &registryFileStore{}
+
+func newRegistryFileManager(a *Allocation, registryFilePath string) *registryFileStore {
+	return &registryFileStore{
+		registryFilePath: registryFilePath,
+		allocation:       a,
+		fileStorer:       a,
+	}
+}
+
+// Update writes the registry file with data provided.
+// This overwrites the current copy stored.
+func (a *registryFileStore) Update(data []byte) error {
+	rf, err := a.getRegistryFile()
+	if err != nil {
+		return errors.New("update_registry_file_failed", "Failed to check existence of registry file: "+err.Error())
+	}
+
+	// if no registry yet, upload new registry file.
+	// otherwise, update registry file.
+	isUpdate := rf != nil
+
+	tempLocal := filepath.Join(os.TempDir(), fmt.Sprintf("registerfile_ul_%d", time.Now().UnixNano()))
+
+	err = os.WriteFile(tempLocal, data, 0600)
+	if err != nil {
+		return errors.New("update_registry_file_failed", "Failed to create file locally for upload: "+err.Error())
+	}
+
+	err = a.fileStorer.StartChunkedUpload(os.TempDir(), tempLocal, a.registryFilePath, nil, isUpdate, false, "", false)
+	if err != nil {
+		return errors.New("update_registry_file_failed", "Failed to upload registry file: "+err.Error())
+	}
+
+	return nil
+}
+
+// Get retrieves the registry file contents and its last updated timestamp.
+func (a *registryFileStore) Get() ([]byte, common.Timestamp, error) {
+	rfLastUpdateTime, err := a.GetLastUpdateTimestamp()
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to check existence of registry file: "+err.Error())
+	}
+
+	// no registry file yet
+	// return empty
+	if rfLastUpdateTime == common.Timestamp(0) {
+		return []byte{}, rfLastUpdateTime, nil
+	}
+
+	// download registry file
+	tempLocal := filepath.Join(os.TempDir(), fmt.Sprintf("starred_dl_%d", time.Now().UnixNano()))
+	cb := newRegistryFileCB()
+
+	err = a.fileStorer.DownloadFile(tempLocal, a.registryFilePath, cb)
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to start download of registry file: "+err.Error())
+	}
+
+	<-cb.Wait() // wait for download to complete
+	if cb.err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to download registry file: "+cb.err.Error())
+	}
+
+	f, err := os.Open(tempLocal)
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to open downloaded registry file: "+err.Error())
+	}
+
+	defer os.Remove(tempLocal)
+	defer f.Close()
+
+	bt, err := io.ReadAll(f)
```

Comment:
ReadAll could cause OOM if the file is large, so perhaps we need to return an io.Reader to caller along with close function to close the reader.

Reply:
Do you see this will be issue for registry file purposes? The idea with @sculptex (at least with starred registry) is the app would keep the list of starred files inmemory. As such, would not expect it to be a lot.

If however we still need to take that large file in consideration, I may need to change not just this download of registry file but also upload. Upload may need to accept an `io.Reader` which expects registry file in correct format already (JSON or lines of paths). Parsing of download into format would be responsibility of caller as well.

Reply:
okay, if the registry file is not too big, we can ignore it atm.